pkg/server/server.go

```diff
@@ -244,10 +244,53 @@ func (s *Server) Run(ctx context.Context) error {
 					VirtualWorkspaceURL: s.Options.Extra.ShardVirtualWorkspaceURL,
 				},
 			}
-			if _, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Create(goContext(ctx), shard, metav1.CreateOptions{}); err != nil {
-				return err
+
+			if err := wait.PollInfiniteWithContext(goContext(ctx), time.Second, func(ctx context.Context) (bool, error) {
+				klog.Infof("Getting %q ClusterWorkspaceShard from the root shard", shard.Name)
+				existingShard, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Get(ctx, shard.Name, metav1.GetOptions{})
+				if err != nil && !errors.IsNotFound(err) {
+					klog.Errorf("failed getting %q ClusterWorkspaceShard from the root shard, err %v", shard.Name, err)
+					return false, nil
+				}
+				if errors.IsNotFound(err) {
+					klog.Info("Creating ClusterWorkspaceShard resource in the root shard because it doesn't exist")
+					if _, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Create(ctx, shard, metav1.CreateOptions{}); err != nil {
+						klog.Errorf("failed creating %q ClusterWorkspaceShard in the root shard, err %v", shard.Name, err)
+						return false, nil
+					}
+					klog.Info("Finished creating ClusterWorkspaceShard resource in the root shard")
+					return true, nil
+				}
+
+				var specChanged bool
+				if len(shard.Spec.BaseURL) > 0 && existingShard.Spec.BaseURL != shard.Spec.BaseURL {
+					existingShard.Spec.BaseURL = shard.Spec.BaseURL
+					specChanged = true
+				}
+				if len(shard.Spec.ExternalURL) > 0 && existingShard.Spec.ExternalURL != shard.Spec.ExternalURL {
+					existingShard.Spec.ExternalURL = shard.Spec.ExternalURL
+					specChanged = true
+				}
+				if len(shard.Spec.VirtualWorkspaceURL) > 0 && existingShard.Spec.VirtualWorkspaceURL != shard.Spec.VirtualWorkspaceURL {
+					existingShard.Spec.VirtualWorkspaceURL = shard.Spec.VirtualWorkspaceURL
+					specChanged = true
+				}
+				if specChanged {
+					_, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Update(ctx, existingShard, metav1.UpdateOptions{})
+					if err != nil {
+						klog.Error("failed updating %q ClusterWorkspaceShard in the root shard, err %v", shard.Name, err)
+						return false, nil
+					}
+					klog.Infof("Updated %q ClusterWorkspaceShard resource in the root shard", shard.Name)
+					return true, nil
+				}
+				klog.Infof("ClusterWorkspaceShard %q resource is up to date in the root shard", shard.Name)
+				return true, nil
```

Comment:
all this worth a method?

Reply:
I'll move it to a separate hook in the next step. So I'd leave it as it is.